pkg/intermediate/priorityqueue.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2021 VMware, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package intermediate
+
+import (
+	"container/heap"
+	"time"
+)
+
+type ItemToExpire struct {
+	// Flow related info
+	flowKey            *FlowKey
+	flowRecord         *AggregationFlowRecord
+	activeExpireTime   time.Time
+	inactiveExpireTime time.Time
+	// Index in the priority queue (heap)
+	index int
+}
+
+type TimeToExpirePriorityQueue []*ItemToExpire
+
+func (pq TimeToExpirePriorityQueue) Len() int {
+	return len(pq)
+}
+
+func (pq TimeToExpirePriorityQueue) Less(i, j int) bool {
+	if pq[i].activeExpireTime.Before(pq[i].inactiveExpireTime) {
+		if pq[j].activeExpireTime.Before(pq[j].inactiveExpireTime) {
+			return pq[i].activeExpireTime.Before(pq[j].activeExpireTime)
+		}
+		return pq[i].activeExpireTime.Before(pq[j].inactiveExpireTime)
+	} else {
+		if pq[j].activeExpireTime.Before(pq[j].inactiveExpireTime) {
+			return pq[i].inactiveExpireTime.Before(pq[j].activeExpireTime)
+		}
+		return pq[i].inactiveExpireTime.Before(pq[j].inactiveExpireTime)
+	}
+}
+
+func (pq TimeToExpirePriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *TimeToExpirePriorityQueue) Push(x interface{}) {
```

Comment:
out of curiosity, why for `Push` and `Pop` we return interface instead of `*ItemToExpire`?

Reply:
These methods are used by the method here https://golang.org/src/container/heap/heap.go?s=1949:1986#L52
Therefore, we are required to pass the interface through this function. We call the Push and Pop functions from the heap go package.
We do not have a complete implementation of the heap here; we piggyback on the heap package.
If the conversions get costly, we may have to move the complete implementation of the heap here instead of using heap go pkg.